subsys/bluetooth/audio/vocs.c

```diff
@@ -0,0 +1,482 @@
+/*  Bluetooth VOCS - Volume offset Control Service
+ *
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/byteorder.h>
+#include <sys/check.h>
+
+#include <device.h>
+#include <init.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+#include <bluetooth/audio/vocs.h>
+
+#include "vocs_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_VOCS)
+#define LOG_MODULE_NAME bt_vocs
+#include "common/log.h"
+
+#define VALID_VOCS_OPCODE(opcode)	((opcode) == VOCS_OPCODE_SET_OFFSET)
+
+#if defined(CONFIG_BT_VOCS)
+static void offset_state_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	BT_DBG("value 0x%04x", value);
+}
+
+static ssize_t read_offset_state(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				 void *buf, uint16_t len, uint16_t offset)
+{
+	struct vocs_server *inst = attr->user_data;
+
+	BT_DBG("offset %d, counter %u", inst->state.offset, inst->state.change_counter);
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &inst->state, sizeof(inst->state));
+}
+
+static void location_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	BT_DBG("value 0x%04x", value);
+}
+
+#endif /* CONFIG_BT_VOCS */
+
+static ssize_t write_location(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			      const void *buf, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	struct vocs_server *inst = attr->user_data;
+	uint32_t old_location = inst->location;
+
+	if (len != sizeof(inst->location)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
```

Comment:
the input argument offset is not checked or used. Should we return invalid offset if it is non-zero?